0000-access-keys.md

```diff
@@ -0,0 +1,313 @@
+/// Key is `account_id`
+struct Account {
+  pub balance: Balance(u128),
+  pub code_hash: Hash,
+  /// Storage usage accounts for all access keys      
+  pub storage_usage: StorageUsage(u64),
+    /// Last block index at which the storage was paid for.
+  pub storage_paid_at: BlockIndex(u64),
+}
+```
+
+### Examples
+
+#### AccessKey as account-level public key
+
+If an AccessKey has the full access to the account and the allowance set to be the max value for u128, then
+it essentially acts as an account-level public key. Which means we can remove account-level
+public keys from the account struct and rely only on access keys.
+
+An access key example from user `vasya.near` with full access:
+```rust
+/// vasya.near,a123bca2
+AccessKey {
+    nonce: 0,
+    
+    permission: AccessKeyPermission::FullAccess,
+}
+```
+
+
+#### AccessKey for a dApp by a user
+
+This is a simple example where a user wants to use some dApp. The user has to authorize this dApp within their wallet, so the dApp knows who the user is, and also can issue simple function call transactions on behalf of this user.
+
+To create such AccessKey a dApp generates a new key pair and passes the new public key to the user's wallet in a URL.
+Then the wallet asks the user to create a new AccessKey with that points to the dApp.
+User has to explicitly confirm this in the wallet for AccessKey to be created.
+
+The new access key is restricted to be only used for the app’s contract_id, but is not restricted for any method name.
+The user also selects the allowance to some reasonable amount, enough for the application to issue regular transactions.
+The application might also hint the user about this desired allowance in some way.
+
+Now the app can issue function call transactions on behalf of the user’s account towards the app’s contract without requiring the user to sign each transaction.
+
+An access key example for chess app from user `vasya.near`:
+```rust
+/// vasya.near,c5d312f3
+AccessKey {
+    nonce: 0,
+    
+    permission: AccessKeyPermission::FunctionCall(FunctionCallPermission {
+        // Since the access key is stored on the Chess app front-end, the user has
+        // limited the spending amount to some reasonable, but large enough number.
+        // NOTE: It's needs to be multiplied to decimals, e.g. 10^-18 
+        allowance: 1_000_000_000,
+        
+        // This access key restricts access to `chess.app` contract.
+        receiver_id: "chess.app",
+
+        // Any method name on the `chess.app` contract can be called.  
+        method_name: None,
+    }),
+}
+```
+
+#### AccessKey issued by a dApp
+
+This is an example where the dApp wants to pay for the user, or it doesn't want to go through the user's sign-in flow.
+For whatever reason the dApp decided to issue an access key directly for their account.
+
+For this to work there should be one account with funds (that dApp controls on the backend) which creates access keys for the users.
```

Comment:
This does not describe where dApp gets public key to generate access key from.

Reply:
Also, the case when user wants to map their existing account to public key I think should be described separately.